proof/refine/ARM/Tcb_R.thy

```diff
@@ -2998,97 +3035,107 @@ lemma tc_corres_sched:
+                      apply (rule corres_option_split; clarsimp?)
+                       apply (rule_tac P=\<top> and P'=\<top> in corres_option_split; clarsimp)
+                       apply (rule schedContextUnbindTCB_corres)
+                      apply (rule corres_when[OF _ schedContextBindTCB_corres], fastforce)
+                     apply (simp only: mapTCBPtr_threadGet get_tcb_obj_ref_def)
+                     apply (rule threadget_corres, clarsimp simp: tcb_relation_def)
+                    apply (wp get_tcb_obj_ref_wp)
+                   apply (wpsimp wp: getTCB_wp simp: mapTCBPtr_def)
+                  apply (rule_tac P=\<top> in hoare_post_taut)
+                 apply (rule hoare_post_taut)
+                apply (rule_tac P=\<top> and P'=\<top> in corres_option_split; clarsimp)
+                apply wpfix
+                apply (rule sp_corres)
+               apply (rule_tac Q="\<lambda>_ s. invs s \<and> valid_sched s \<and> simple_sched_action s \<and>
+                                        tcb_at t s \<and> ex_nonz_cap_to t s \<and>
+                                        (\<forall>scPtr. sc_opt = Some (Some scPtr) \<longrightarrow>
+                                                   ex_nonz_cap_to scPtr s \<and>
+                                                   sc_tcb_sc_at ((=) None) scPtr s \<and>
+                                                   bound_sc_tcb_at ((=) None) t s)"
+                            in hoare_strengthen_post[rotated])
+                apply (clarsimp simp: obj_at_def split: option.splits)
+                apply (frule invs_valid_objs)
+                apply (erule (1) valid_objsE)
+                apply (clarsimp simp: valid_obj_def valid_tcb_def obj_at_def is_sc_obj_def
+                                      invs_def valid_state_def valid_pspace_def)
+                apply (clarsimp split: Structures_A.kernel_object.splits)
+                apply (drule (2) bound_sc_tcb_bound_sc_at[where tcbptr=t])
+                 apply (clarsimp simp: pred_tcb_at_def obj_at_def)
+                apply (clarsimp simp: sc_at_ppred_def obj_at_def)
+               apply (wpsimp wp: set_priority_valid_sched hoare_vcg_all_lift hoare_vcg_const_imp_lift)
+              apply (rule_tac Q=" \<lambda>_ s. invs' s \<and> tcb_at' t s \<and> ex_nonz_cap_to' t s \<and>
+                                        (\<forall>scp. sc_opt = Some (Some scp) \<longrightarrow> ex_nonz_cap_to' scp s)"
+                           in hoare_strengthen_post[rotated], fastforce split: option.split)
+              apply (wpsimp wp: setP_invs' hoare_vcg_all_lift hoare_vcg_const_imp_lift)
+             apply clarsimp
+             apply (prop_tac "ct_active s \<and> ct_released s", erule conjunct1)
+             apply simp
+            apply clarsimp
+            apply (prop_tac "invs' s' \<and> tcb_at' t s' \<and> ex_nonz_cap_to' t s' \<and>
```

Comment:
I'm slightly confused about prop_tac --> assumption. If assumption proves the statement, doesn't that mean the statement already exists in the hypotheses, and so the prop_tac is not useful? Can you explain briefly what's happening here?

Reply:
There are two `stronger_corres_guard_imp` applications in the proof, and we resolve the concrete guard of one of them at this point. `prop_tac` is used to introduce the stronger guard, then `assumption` is used to unify it with the appropriate schematic.